Multi-server: Enhancing sched to work with partition tolerance, 1

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
This enhances scheduler to work even when one/more servers down in a multi-server setup. This is first of a multi-part change. Here, we are enhancing sched to not stall scheduling when one/more servers are down. This doesn't handle multi-node job spanning server boundaries, this will be handled in further PRs.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->


#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
```
[root@pbspro ravi]# PBS_CONF_FILE=/etc/pbs_1.conf /etc/init.d/pbs start
Starting PBS
/opt/pbs/sbin/pbs_comm ready (pid=83015), Proxy Name:pbspro:17001, Threads:4
PBS comm
PBS mom
PBS sched
Connecting to PBS dataservice...connected to PBS dataservice@pbspro
PBS server
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:15001 qsub -- /bin/sleep 10000
1400.pbspro
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
1200.pbspro       STDIN            ravi              00:00:00 R workq           
501.pbspro        STDIN            ravi              00:00:00 R workq           
1300.pbspro       STDIN            ravi              00:00:00 R workq           
1400.pbspro       STDIN            ravi                     0 R workq           
[ravi@pbspro ~]$ qdel 1200.pbspro 1300.pbspro
qdel: cannot connect to server pbspro runs on port 18000 (errno=111)
[root@pbspro ravi]# PBS_CONF_FILE=/etc/pbs_2.conf /etc/init.d/pbs start
Starting PBS
Connecting to PBS dataservice...connected to PBS dataservice@pbspro
PBS server
[root@pbspro ravi]# 
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:18000 qsub -- /bin/sleep 10000
601.pbspro
[ravi@pbspro ~]$ sudo /opt/pbs/bin/qmgr -c 's s scheduling=t'
[ravi@pbspro ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
501.pbspro        STDIN            ravi              00:00:00 R workq           
1400.pbspro       STDIN            ravi              00:00:00 R workq           
601.pbspro        STDIN            ravi                     0 R workq           
[ravi@pbspro ~]$ ps -ef | grep pbs_server
root     83182     1  0 02:07 ?        00:00:00 /opt/pbs/sbin/pbs_server.bin
root     83435     1  0 02:08 ?        00:00:00 /opt/pbs/sbin/pbs_server.bin
ravi     83477    36  0 02:10 pts/0    00:00:00 grep --color=auto pbs_server
[ravi@pbspro ~]$ sudo kill 83435
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=15031)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
501.pbspro        STDIN            ravi              00:00:00 R workq           
1400.pbspro       STDIN            ravi              00:00:00 R workq           
601.pbspro        STDIN            ravi              00:00:00 R workq           
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:15001 qsub -- /bin/sleep 10000
1500.pbspro
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
501.pbspro        STDIN            ravi              00:00:00 R workq           
1400.pbspro       STDIN            ravi              00:00:00 R workq           
601.pbspro        STDIN            ravi              00:00:00 R workq           
1500.pbspro       STDIN            ravi                     0 Q workq  
[ravi@pbspro ~]$ sudo /opt/pbs/bin/qmgr -c 's s scheduling=t'
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
501.pbspro        STDIN            ravi              00:00:00 R workq           
1400.pbspro       STDIN            ravi              00:00:00 R workq           
601.pbspro        STDIN            ravi              00:00:00 R workq           
1500.pbspro       STDIN            ravi                     0 R workq 
```
**Preemption with 1st server down:**
```
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:15001 qsub -- /bin/sleep 1000
2600.pbspro
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
2600.pbspro       STDIN            ravi              00:00:00 R workq           
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:15001 qsub -lncpus=3 -- /bin/sleep 1000
2700.pbspro
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
2600.pbspro       STDIN            ravi              00:00:00 R workq           
2700.pbspro       STDIN            ravi                     0 R workq           
[ravi@pbspro ~]$ PBS_SERVER_INSTANCES=pbspro:15001 qsub -q highp -- /bin/sleep 1000
2800.pbspro
[ravi@pbspro ~]$ qstat
qstat: cannot connect to server pbspro runs on port 18000 (errno=111)
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
2600.pbspro       STDIN            ravi              00:00:00 R workq           
2700.pbspro       STDIN            ravi              00:00:00 S workq           
2800.pbspro       STDIN            ravi                     0 R highp     
```

**preemption with both servers up, preempting jobs from both**:
```
[ragrawal@m1 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
900.m1            STDIN            ragrawal          00:00:00 R workq           
1000.m1           STDIN            ragrawal          00:00:00 R workq           
501.m1            STDIN            ragrawal          00:00:00 R workq           
601.m1            STDIN            ragrawal          00:00:00 R workq           
[ragrawal@m1 ~]$ qsub -q highp -lselect=2:ncpus=4 -- /bin/sleep 100
1100.m1
[ragrawal@m1 ~]$ qstat
Job id            Name             User              Time Use S Queue
----------------  ---------------- ----------------  -------- - -----
900.m1            STDIN            ragrawal          00:00:00 S workq           
1000.m1           STDIN            ragrawal          00:00:00 S workq           
1100.m1           STDIN            ragrawal                 0 R highp           
501.m1            STDIN            ragrawal          00:00:00 S workq           
601.m1            STDIN            ragrawal          00:00:00 S workq 
```


Regression Results:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7696|4178|0|0|0|15|4163|

<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

nithinj commented on src/include/libpbs.h:
```diff
@@ -414,6 +412,9 @@ int get_server_fd_from_jid(int c, char *jobid);
 int multi_svr_op(int fd);
 int get_job_svr_inst_id(int c, char *job_id);
 
+int pbs_register_sched_single_msvr(const char *sched_id, int primary_conn_id, int secondary_conn_id);
```
why single and msvr both appear in the name?

nithinj commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -618,6 +621,43 @@ connect_to_servers(char *svrhost, uint port, char *extend_data)
+ * @return void
+ */
+void
+pbs_connect_single_msvr(svr_conn_t *conn)
```
can you rename it to pbs_connect_msvr_instance or similar?

nithinj commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -618,6 +621,43 @@ connect_to_servers(char *svrhost, uint port, char *extend_data)
+ * @retval	-1 for error
+ */
+int
+pbs_disconnect_single_msvr(svr_conn_t *svr_conn)
```
can you rename it to pbs_disconnect_msvr_instance or similar?

arungrover commented on src/scheduler/pbs_sched_utils.cpp:
```diff
@@ -684,13 +744,56 @@ sched_svr_init(void)
+	part_tolerance = true;
+
+	if (time(NULL) - last_reconnect < min_reconnect_period) {
+		return;	/* To throttle the reconnect frequency, don't connect if under min_reconnect_period */
```
Unlock the mutex. We should consider making a "Locker" class that takes a given mutex in the constructor and releases it in the destructor. That way we don't have to worry about forgetting to release mutexes.

agrawalravi90 replied:
Great idea, but as we discussed, let's wait until we refactor code to use std:thread and std:mutex, then we can use STL's std::lock_guard or std::unique_lock to do this.

arungrover commented on src/scheduler/pbs_sched_utils.cpp:
```diff
@@ -684,13 +744,56 @@ sched_svr_init(void)
+		return;
+	}
+
+	part_tolerance = true;
```
Shouldn't we be setting partition_tolerance after we have checked min_reconnect_period? Can it happen that reconnect_servers is called frequently twice and it makes the scheduler go into partition tolerance mode even when no server was down?

agrawalravi90 replied:
reconnect_servers() should only be called when one/more servers is down, so I wanted to set part_tolerance upfront .. but I'll move it to the if condition that checks how many are down ..

arungrover commented on src/scheduler/pbs_sched_utils.cpp:
```diff
@@ -684,13 +744,56 @@ sched_svr_init(void)
 	pthread_mutex_lock(&cleanup_lock);
+	if (prim_conn == NULL || sec_conn == NULL || get_num_servers() == 1) {
+		close_servers();
+		connect_svrpool();
```
Since connect_svrpool can also set part_tolerance to True, I feel we should always set it to False here if the number of servers configured is only 1.

agrawalravi90 replied:
connect_svrpool() sets it to False the first thing, so we should be ok.

arungrover commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -153,7 +177,11 @@ send_attr_updates(int virtual_sd, resource_resv *resresv, struct attrl *pattr)
-    return pbs_preempt_jobs(virtual_sd, preempt_jobs_list);
+	preempt_job_info *ret;
+
+    ret = pbs_preempt_jobs(virtual_sd, preempt_jobs_list);
```
As of today, pbs_preempt_jobs already detects if it is not able to read data from one of the servers it clears the responses from other servers too. Are you going to address that in later PRs?

agrawalravi90 replied:
Thanks for pointing this out, I hadn't thought about it, yes, it'll need to be updated.

arungrover commented on src/scheduler/sched_ifl_wrappers.cpp:
```diff
@@ -153,7 +177,11 @@ send_attr_updates(int virtual_sd, resource_resv *resresv, struct attrl *pattr)
+
+    ret = pbs_preempt_jobs(virtual_sd, preempt_jobs_list);
+
+	return static_cast<preempt_job_info *>(handle_part_tolerance(ret));
```
This can potentially become a leak if upon a sigpipe this function returns 0 and does not free up responses in "ret" variable.
It can also happen in all other stat functions below.

agrawalravi90 commented on src/scheduler/pbs_sched_utils.cpp:
```diff
@@ -534,6 +534,60 @@ are_we_primary()
+		return 1;
+	}
+
+	/* Both primary and secondary successful, add secondary to poll list */
```
Actually this means to say that both were successfully connected, so we can add secondary to the poll list. I'll make it more clear.

arungrover replied:
ah, no worries, I misread it.

agrawalravi90 commented on src/scheduler/pbs_sched_utils.cpp:
```diff
@@ -610,28 +666,30 @@ connect_svrpool()
-				break;
+			    svr_conns_secondary[i]->state == SVR_CONN_STATE_DOWN) {
+				/* At least one server is down, operate in partition tolerance mode */
+				part_tolerance = true;
```
this is not secondary server, it's the secondary connection to the same server. I think it's safer to disconnect both primary and secondary connections when either of them couldn't be connected, so I think I'll make that change. We will try reconnecting both connections after every 30 seconds, so it should be ok to be conservative. Let me know what you think.

arungrover replied:
Sorry, I again misread it. If it is the secondary connection then you are right. There is no point in keeping partially connected server, better to reconnect both connections.

agrawalravi90 commented:
@arungrover  and @nithinj  thanks for your review, I've addressed your comments. I also found a bug in preemption with multi-server, we were not updating svr_to_psets in update_all_nodepart, which was causing preemption to fail as the node partition could show free_nodes=0 even when some nodes were free'd up by the simulation. Please check it out and let me know if it looks okay.

arungrover commented on src/lib/Libifl/pbsD_Preempt_Jobs.c:
```diff
@@ -156,9 +155,10 @@ __pbs_preempt_jobs(int c, char **preempt_jobs_list)
+	int retidx = -1;	/* first server's list will be used to collate results from all */
 	preempt_job_info *ret = NULL;
 	void *missing_jobs = NULL;
+	int num_servers = get_num_servers();
```
you can use num_servers on line 166 as well

arungrover reviewed (approved):
I just have this minor comment. LGTM!

agrawalravi90 commented:
@arungrover  thanks! I've made a fix to server local scheduling, which was using psets and seeing that a superchunk was split across nodes of different hosts because allpart had the ok_break bit set. I changed it to be set to false by default, which fixed the issue. Please look at the latest changes.

@nithinj please provide further feedback

nithinj commented on src/lib/Libifl/pbsD_Preempt_Jobs.c:
```diff
@@ -87,7 +87,8 @@ preempt_jobs_send(int connect, char **preempt_jobs_list)
 
 	if (dis_flush(connect)) {
-		pbs_errno = PBSE_PROTOCOL;
+		if (pbs_errno == PBSE_NONE)
```
why checking for PBSE_NONE? I could not see this check with any of the errors above.

nithinj replied:
Does dis layer set pbs error code? I don't think so. but if that's the case you might want to add the check to other places in the function.

nithinj reviewed (commented):
Rest looks good to me!

nithinj reviewed (approved):
LGTM!
